src/Nethermind/Nethermind.Trie/NodeStorage.cs

```diff
@@ -0,0 +1,211 @@
+    {
+        Debug.Assert(pathSpan.Length == StoragePathLength);
+
+        // Key structure look like this.
+        //
+        // For state (total 42 byte)
+        //
+        // +--------------+------------------+------------------+--------------+
+        // | section byte | 8 byte from path | path length byte | 32 byte hash |
+        // +--------------+------------------+------------------+--------------+
+        //
+        // For storage (total 74 byte)
+        // +--------------+---------------------+------------------+------------------+--------------+
+        // | section byte | 32 byte from address | 8 byte from path | path length byte | 32 byte hash |
+        // +--------------+---------------------+------------------+------------------+--------------+
+        //
+        // The section byte is:
+        // - 0 if state and path length is <= 5.
+        // - 1 if state and path length is > 5.
+        // - 2 if storage.
+        //
+        // The keys are separated due to the different characteristics of these nodes. The idea being that top level
+        // node can be up to 5 times bigger than lower node, and grew a lot due to pruning. So mixing them makes lower
+        // node sparser and have poorer cache hit, and make traversing leaves for snap serving slower.
+        //
+        // Technically, you'll need 9 byte for state and 8 byte for storage on mainnet for the path. But we want to keep
+        // key small at the same time too. If the key are too small, multiple node will be out of order, which
+        // can be slower but as long as they are in the same data block, it should not make a difference.
+        // On mainnet, the out of order key is around 0.03% for address and 0.07% for storage.
+
+        if (address == null)
+        {
+            // Separate the top level tree into its own section. This marginally improve cache hit rate, but not much.
+            // 70% of duplicated keys is in this section, making them pretty bad, so we isolate them here to not expand
+            // the space of other things, hopefully we can cache them by key somehow. Separating by the path length 4
+            // does improve cache hit and processing time a little bit, until a few hundreds prune persist where it grew
+            // beyond block cache size.
+            if (path.Length <= 5)
+            {
+                pathSpan[0] = 0;
+            }
+            else
+            {
+                pathSpan[0] = 1;
+            }
+
+            // Keep key small
+            path.Path.BytesAsSpan[..8].CopyTo(pathSpan[1..]);
+            keccak.Bytes.CopyTo(pathSpan[10..]);
+
+            pathSpan[9] = (byte)path.Length;
+            return pathSpan[..42];
+        }
+        else
+        {
+            pathSpan[0] = 2;
+            address.Bytes.CopyTo(pathSpan[1..]);
+            path.Path.BytesAsSpan[..8].CopyTo(pathSpan[33..]);
```

Comment:
Address is only 20 bytes, why use 32 bytes for it?

Reply:
Its the hash of the address here. So its 32 byte.

Reply:
Any reason not to just use the address? Would save 12 bytes per key and should be equivlent in exactness (or even more precise as its the actual address)

Reply:
Hard to pass the address through the patricia trie. Need a bit more refactor to work, notably the call recursive thingy.

Reply:
Technically, it would still work with even 10byte. But its the full 32 byte to work with #6439 . I guess 16 byte prefix would work also as long as it does not collide with other storage, or resistent to collision to prevent attack.

Reply:
Check snap sync protcol, you are right it uses address hash so this is the better format